pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -0,0 +1,253 @@
+		).ToController("ClusterMemberRemovalController", eventRecorder.WithComponentSuffix("cluster-member-removal-controller"))
+}
+
+func (c *ClusterMemberRemovalController) sync(ctx context.Context, _ factory.SyncContext) error {
+	if !c.hasNetworkInformerSyncedFn() {
+		return nil
+	}
+	// stop if the machine API is not functional
+	if isFunctional, err := c.isMachineAPIFunctionalFn(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	etcdEndpointsConfigMap, err := c.configMapListerForTargetNamespace.Get("etcd-endpoints")
+	if err != nil {
+		return err // should not happen
+	}
+	if len(etcdEndpointsConfigMap.Data) == 0 {
+		return nil
+	}
+	for _, potentialMemberToRemoveIP := range etcdEndpointsConfigMap.Data {
+		memberLocator := potentialMemberToRemoveIP
+		machine, err := c.getMachineForMember(potentialMemberToRemoveIP)
+		if err != nil {
+			klog.Errorf("unable to get a machine for member: %v, err: %v", memberLocator, err)
+			continue
+		}
+		if machine != nil {
+			klog.V(4).Infof("cannot remove member: %v because a machine resource still exists (%v/%v)", memberLocator, machine.Name, machine.UID)
+			continue
+		}
+
+		node, err := c.getNodeForMember(potentialMemberToRemoveIP)
+		if err != nil {
+			klog.Errorf("unable to get a node for member: %v, err: %v", memberLocator, err)
+			continue
+		}
+		if node != nil {
+			klog.V(4).Infof("cannot remove member: %v because a node resource still exists (%v/%v)", memberLocator, node.Name, node.UID)
+			continue
+		}
+
+		// it looks like we have found a member that isn't backed by a machine nor a node
+		// issue a live request to the etcd cluster and remove it from the cluster
+		memberList, err := c.etcdClient.MemberList(ctx)
+		if err != nil {
+			return err
+		}
+		for _, member := range memberList {
+			memberIP, err := memberToNodeInternalIP(member)
+			if err != nil {
+				return err
+			}
+			if memberIP != potentialMemberToRemoveIP {
+				continue
+			}
+			memberLocator = fmt.Sprintf("[ url: %v, name: %v, id: %v ]", memberIP, member.Name, member.ID)
+			if len(member.Name) == 0 {
+				// TODO: what if the member hasn't been started and the name field is empty ?!
+				//       we are going to change the MemberRemove method, see L145
+				klog.Errorf("unable to remove member: %v because it doesn't have a name (has the member been started?)", memberLocator)
+				break
+			}
+			// TODO (polynomial) change the signature of the MemberRemove method to accept an ID not a Name.
+			err = c.etcdClient.MemberRemove(ctx, member.Name)
+			if err != nil {
+				klog.Errorf("failed to remove member: %v, err: %v", memberLocator, err)
+			} else {
+				klog.V(2).Infof("successfully removed member: %v from the cluster", memberLocator)
+			}
+		}
+	}
+	return nil
+}
+
+func (c *ClusterMemberRemovalController) getMachineForMember(memberInternalIP string) (*machinev1beta1.Machine, error) {
```

Comment:
Add doc here explaining that not-found is not an Error like it is for "normal" clients